pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -1091,13 +1087,38 @@ func BuildDefaultHTTPInboundRoute(clusterName string, operation string) *route.R
 	return val
 }
 
+// setTimeout sets timeout for a route.
+func setTimeout(action *route.RouteAction, vsTimeout *duration.Duration, node *model.Proxy) {
+	// Configure timeouts specified by Virtual Service if they are provided, otherwise set it to defaults.
+	action.Timeout = features.DefaultRequestTimeout
+	if vsTimeout != nil {
+		action.Timeout = vsTimeout
+	}
+	action.MaxStreamDuration = &route.RouteAction_MaxStreamDuration{}
+	if node != nil && node.IsProxylessGrpc() {
+		// TODO(stevenctl) merge these paths; grpc's xDS impl will not read the deprecated value
+		action.MaxStreamDuration.MaxStreamDuration = action.Timeout
+	} else {
+		// Set MaxStreamDuration only for notimeout cases otherwise it wont be honored.
+		if action.Timeout.AsDuration().Nanoseconds() == 0 {
+			action.MaxStreamDuration.MaxStreamDuration = action.Timeout
+			action.MaxStreamDuration.GrpcTimeoutHeaderMax = action.Timeout
+		} else {
+			// If not configured at all, the grpc-timeout header is not used and
+			// gRPC requests time out like any other requests using timeout or its default.
+			// Use deprecated value for now as the replacement MaxStreamDuration has some regressions.
+			// nolint: staticcheck
+			action.MaxGrpcTimeout = action.Timeout
```

Comment:
I have a small quick question about this line. Are MaxStreamDuration and MaxGrpcTimeout exactly the same? There is also a MaxStreamDuration field in my project, but without MaxGrpcTimeout.

Reply:
They are not exactly same. Please see https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html?highlight=max_grpc_timeout#config-route-v3-routeaction-maxstreamduration

There are issues with max_stream_duration. Refer to https://github.com/envoyproxy/envoy/issues/19073 and https://github.com/envoyproxy/envoy/issues/19073